draft-ietf-tls-esni.md

```diff
@@ -994,6 +974,72 @@ The "ech_is_inner" is expected to have an empty payload. If the payload is
 non-empty (i.e., the length of the "extension_data" field is non-zero) then the
 backend server MUST abort the handshake with an "illegal_parameter" alert.
 
+## Sending HelloRetryRequest {#server-hrr}
+
+If the backend server cannot complete the handshake with the parameters provided
+in the ClientHelloInner, then it constructs a HelloRetryRequestInner message in
+the usual way and forwards it to the client-facing server. The client-facing
+server encrypts this message using a key derived from the HPKE context, as
+described below. It then constructs HelloRetryRequestOuter message with an
+"encrypted_client_hello" extension containing the ciphertext as its payload. It
+then continues the handshake by sending this message to the client.
+
+The payload of the "encrypted_client_hello" extension is constructed as follows.
+The client-facing server begins by computing `hrr_inner_key` and
+`hrr_inner_nonce` as described in {{client-hrr-accepted-ech}}. The extension
```

Comment:
(This is a note not just about this PR but the document in general.) I'm finding it confusing for all the client behavior discussed in one section and the server behavior discussed in another. I'm bouncing back and forth from one to the other to get the actual ordering of the series of events. Would it not be less confusing to put things in that order than to describe all the client behavior and then all the server behavior?

Reply:
Hmmm, that's an interesting idea. I wonder which is easier from the perspective of someone who is implementing this for the first time. My assumption has been that it's better to separate client behavior and server behavior into two section so that the implementer needs only to refer to one section when implementing the client (resp. server) bits. But given how complex ECH is, perhaps a chronological description is better. I think this sort of editorial re-work is worth filing an issue for discussion.